README
App-DBBrowser
=============



INFO

Search and read in SQLite/MySQL/PostgreSQL databases.



INSTALLATION

    perl Makefile.pl
    make
    make test
    make install



HELP

db-browser -h




LICENSE AND COPYRIGHT

Copyright (C) 2012-2014 Matthäus Kiem

This program is free software; you
can redistribute it and/or modify it under the same terms
as Perl 5.10.0.

See http://dev.perl.org/licenses/ for more information.

This program is distributed in the hope that it will be
useful, but without any warranty; without even the implied
warranty of merchantability or fitness for a particular purpose.


Changes
Revision history for App-DBBrowser


0.048   2014-11-20
        - Update documentation.
        - "Term::Choose" minimum version "1.117".

0.047   2014-11-11
        - New option "Debug".
        - Code refactoring.
        - Update documentation.

0.046_01   2014-11-08
        - Added file history.
        - Choose parse module: Textfiles instead of only CSV files.
        - Update documentation.

0.046   2014-11-07
        - Support file formats supported by "Spreadsheet::Read".
        - New dependencies: "Spreadsheet::Read" and "Text::ParseWords".
        - Code refactoring.
        - Update documentation.

0.045   2014-10-29
        - Code refactoring.

0.044_03   2014-10-28
        - Switched back to "Text::CSV".
        - Added missing dependency in the "Makefile.PL".
        - Bugfix.
        - Update documentation.

0.044_02   2014-10-28
        - Refactoring input.

0.044_01   2014-10-27
        - Bugfixes input.

0.044   2014-10-27
        - Use of "Text::ParseWords" instead of "Text::CSV" to parse rows.
        - Update documentation.

0.043   2014-10-02
        - Bugfix "print_error_message" [github #2 - Mark Lawrence].
        - Don't print all filenames when searching SQLite databases.

0.042   2014-10-01
        - Handle undefined input filename.

0.041_02   2014-09-29
        - Insert: additional option to choose single rows.
        - Modified menu behavior.

0.041_01   2014-09-27
        - Let "SET" sub-statement behave like similar other sql sub-statements.

0.041   2014-09-26
        - Code cleanup.
        - Improved documentation.

0.040_05   2014-09-23
        - Bugfix row-col filter.
        - Bugfix temp csv config.
        - Update menu navigation.

0.040_04   2014-09-23
        - Update menu navigation.
        - Update documentation.

0.040_03   2014-09-23
        - Update "INSERT INTO".
        - Code refactoring.
        - Update documentation.

0.040_02   2014-09-20
        - Bugfix "print_error_message".

0.040_01   2014-09-19
        - Added experimental sql features (update, delete, insert into).
        - Update documentation.

0.040   2014-09-03
        - Menu memory for config and SQL menus.
        - Replaced "Term::ReadLine::Tiny" with "Term::ReadLine::Simple".
        - Update documentation.

0.039   2014-08-16
        - Perl minimum version from "5.10.1" to "5.10.0".

0.038   2014-08-16
        - Improved documentation.

0.037   2014-07-28
        - Bugfix: regex case sensitivity setting.
        - New dependency: "Term::ReadLine::Tiny".
        - "use JSON" instead of "use JSON::XS".
        - Update Makefile.PL.
        - Code refactoring.
        - Update documentation.

0.036   2014-07-10
        - Required minimum version of "Term::Choose::Util" is now "0.011".
        - Update documentation.

0.035_04   2014-07-09
        - Required minimum version of "Term::Choose::Util" is now "0.010".
        - Some read line changes.
        - Update documentation.

0.035_03   2014-07-08
        - Bugfix in JSON encoding.
        - Update "Reset".
        - Code refactoring.
        - Update documentation.

0.035_02   2014-07-05
        - Code refactoring.
        - Fixed option "HELP".

0.035_01   2014-07-04
        - New configuration file format: each database driver has now its own file.
        - One option ("Menus") instead of three to set table/schema/database menu memory behavior.
        - The fetching of the required data for connecting to databases has changed.
        - Changed "Reset" option.
        - Required minimum version of "Term::TablePrint" is now "0.013".
        - Required minimum version of "Term::Choose::Util" is now "0.009".

0.035   2014-06-29
        - Required minimum version of "Term::Choose::Util" is now "0.008".
        - New option to set the default directories where "db-browser" searches for SQLite databases.
        - Update documentation.

0.034   2014-06-18
        - Bugfix in LIMIT-submenu.
        - Update documentation.

0.033   2014-06-15
        - Raised the required minimum version of "Term::TablePrint" and "Term::Choose::Util".
        - Changed the layout of the "Path" output.
        - Fixed: a change of the setting of the option "mouse" is immediately effective.

0.032   2014-05-30
        - Printing the table header on each page is not enabled by default.

0.031   2014-05-28
        - Fixed bug in the option code.

0.030   2014-05-28
        - Added an option to reset the database specific settings back to "Database defaults".
        - Code refactoring.
        - Update documentation.

0.029_02   2014-05-25
        - Fixed option "Binary Filter".
        - Removed the extra code which moved "YES" before "NO".
        - Code refactoring.
        - Update documentation.

0.029_01   2014-05-21
        - Each package has now its own file.
        - Renamed the option "Expand" to "Enchant"
        - Added the option "Table Header" in "Enchant".
        - Update documentation.

0.029   2014-05-05
        - Parentheses: the user can choose if the parentheses menu entry is located before or after the column names.
        - Update documentation.

0.028   2014-04-28
        - Bugfix documentation.

0.027   2014-04-28
        - Improved documentation.

0.026   2014-04-28
        - Improved documentation.

0.025   2014-04-26
        - SQL menu entry "COLUMNS" is now called "SELECT".
        - Code refactoring.
        - Improved documentation.

0.024   2014-04-21
        - Bugfix in "backup_in_hidd".
        - It is possible now also with "GROUP BY" to choose more than one column at once.
        - Update documentation.

0.023   2014-04-10
        - Update documentation.

0.022   2014-03-24
        - Update "ALL TABLES" in UNION.
        - Mininum version of "Term::TablePrint" is now 0.005.
        - Update Makefile.PL

0.021   2014-03-22
        - Bugfix in UNION.
        - New: "ALL TABLES" in UNION.

0.020   2014-03-21
        - Rewritten in OO style.
        - Outsourced code to "Term::Choose::Util".


Makefile.PL
use 5.010001;
use warnings;
use strict;
use ExtUtils::MakeMaker;


my %prereqs_os_specific;
if ( $^O eq 'MSWin32' ) {
    %prereqs_os_specific = (
        'Win32::Console::ANSI' => 0,
    );
}


WriteMakefile(
    PL_FILES         => {},
    MIN_PERL_VERSION => '5.010001',
    LICENSE          => 'perl',
    META_MERGE => {
        'meta-spec' => { version => 2 },
        resources   => {
            repository => {
              type => 'git',
              web  => 'https://github.com/kuerbis/App-DBBrowser',
              url  => 'https://github.com/kuerbis/App-DBBrowser.git',
            },
        },
    },
    AUTHOR           => 'Matthaeus Kiem <cuer2s@gmail.com>',
    NAME             => 'App::DBBrowser',
    ABSTRACT_FROM    => 'lib/App/DBBrowser.pm',
    VERSION_FROM     => 'lib/App/DBBrowser.pm',
    EXE_FILES        => [ 'bin/db-browser' ],
    BUILD_REQUIRES   => {
        'Test::More'            => 0,
    },
    PREREQ_PM => {
        'Clone'                 => 0,
        'DBI'                   => 0,
        'Encode'                => 0,
        'Encode::Locale'        => 0,
        'File::Basename'        => 0,
        'File::Find'            => 0,
        'File::HomeDir'         => 0,
        'File::Spec::Functions' => 0,
        'FindBin'               => 0,
        'Getopt::Long'          => 0,
        'JSON::XS'              => 0,
        'List::MoreUtils'       => 0,
        'Pod::Usage'            => 0,
        'Scalar::Util'          => 0,
        'Term::Choose'          => '1.105',
        'Term::Choose::Util'    => 0,
        'Term::TablePrint'      => '0.010_01',
        'Text::LineFold'        => 0,
        %prereqs_os_specific,
    },
    dist  => {
        COMPRESS => 'gzip',
        SUFFIX   => '.gz',
        TARFLAGS => '--format=gnu -cvf',
    },
    clean => {
        FILES => 'App-DBBrowser-*',
    },
);


lib/App/DBBrowser/Opt.pm
package # hide from PAUSE
App::DBBrowser::Opt;

use warnings;
use strict;
use 5.010001;

our $VERSION = '0.033';

use Encode                qw( encode );
use File::Basename        qw( basename );
use File::Spec::Functions qw( catfile );
use FindBin               qw( $RealBin $RealScript );
#use Pod::Usage            qw( pod2usage );             # "require"-d in options/help

use Clone              qw( clone );
use Encode::Locale     qw();
use JSON::XS           qw( decode_json );
use Term::Choose       qw( choose );
use Term::Choose::Util qw( insert_sep print_hash util_readline choose_a_number choose_a_subset choose_multi );

sub new {
    my ( $class, $info, $opt ) = @_;
    bless { info => $info, opt => $opt }, $class;
}


sub options {
    my ( $self ) = @_;
    my $menus = [
        [ 'db_defaults',    "- DB Defaults" ],
        [ 'db_drivers',     "- DB Drivers" ],
        [ 'db_login_once',  "- DB Login" ],
        [ '_env_dbi',       "- ENV DBI" ],
        [ '_enchant',       "- Enchant" ],
        [ 'lock_stmt',      "- Lock" ],
        [ 'max_rows',       "- Max Rows" ],
        [ 'metadata',       "- Metadata" ],
        [ 'mouse',          "- Mouse Mode" ],
        [ 'min_col_width',  "- Colwidth" ],
        [ 'operators',      "- Operators" ],
        [ '_parentheses',   "- Parentheses" ],
        [ 'progress_bar',   "- ProgressBar" ],
        [ 'regexp_case',    "- Regexp Case" ],
        [ 'sssc_mode',      "- Sssc Mode" ],
        [ 'tab_width',      "- Tabwidth" ],
        [ 'undef',          "- Undef" ],
    ];
    my $sub_menus = {
       _enchant      => [
            [ 'keep_db_choice',     "- Choose Database", [ 'Simple', 'Memory' ] ],
            [ 'keep_schema_choice', "- Choose Schema",   [ 'Simple', 'Memory' ] ],
            [ 'keep_table_choice',  "- Choose Table",    [ 'Simple', 'Memory' ] ],
            [ 'table_expand',       "- Print  Table",    [ 'Simple', 'Expand' ] ],
            [ 'keep_header',        "- Table Header",    [ 'Simple', 'Each page' ] ],
        ],
        _parentheses => [
            [ 'w_parentheses', "- Parentheses in WHERE",     [ 'NO', '(YES', 'YES(' ] ],
            [ 'h_parentheses', "- Parentheses in HAVING TO", [ 'NO', '(YES', 'YES(' ] ],
        ],
        _env_dbi     => [
            [ 'env_dbi_user', "- Use DBI_USER", [ 'NO', 'YES' ] ],
            [ 'env_dbi_pass', "- Use DBI_PASS", [ 'NO', 'YES' ] ],
        ],
    };
    my $path = '  Path';
    my @pre = ( undef, $self->{info}{_continue}, $self->{info}{_help}, $path );
    my @real = map( $_->[1], @$menus );
    my $choices = [ @pre, @real ];

    OPTION: while ( 1 ) {
        # Choose
        my $idx = choose(
            $choices,
            { %{$self->{info}{lyt_3}}, index => 1, undef => $self->{info}{_exit} }
        );
        exit if ! defined $idx;
        my $key;
        if ( $idx <= $#pre ) {
            $key = $pre[$idx];
        }
        else {
            $idx -= @pre;
            $key = $menus->[$idx][0];
            die if $key !~ /^_/ && ! exists $self->{opt}{$key};
        }
        if ( ! defined $key ) {
            if ( $self->{info}{write_config} ) {
                $self->__write_config_file( $self->{info}{config_file} );
                delete $self->{info}{write_config};
            }
            exit();
        }
        elsif ( $key eq $self->{info}{_continue} ) {
            if ( $self->{info}{write_config} ) {
                $self->__write_config_file( $self->{info}{config_file} );
                delete $self->{info}{write_config};
            }
            return $self->{opt};
        }
        elsif ( $key eq $self->{info}{_help} ) {
            require Pod::Usage;
            Pod::Usage::pod2usage( {
                -input => 'bin/db-browser',
                -exitval => 'NOEXIT',
                -verbose => 2 } );
        }
        elsif ( $key eq $path ) {
            my $version = 'version';
            my $bin     = '  bin  ';
            my $app_dir = 'app-dir';
            my $path = {
                $version => $main::VERSION,
                $bin     => catfile( $RealBin, $RealScript ),
                $app_dir => $self->{info}{app_dir},
            };
            my $keys = [ $version, $bin, $app_dir ];
            print_hash( $path, { keys => $keys, preface => ' Close with ENTER' } );
        }
        elsif ( $key eq 'tab_width' ) {
            my $digits = 3;
            my $prompt = 'Tab width';
            $self->__opt_number_range( $key, $prompt, $digits );
        }
        elsif ( $key eq 'min_col_width' ) {
            my $digits = 3;
            my $prompt = 'Minimum Column width';
            $self->__opt_number_range( $key, $prompt, $digits );
        }
        elsif ( $key eq 'undef' ) {
            my $prompt = 'Print replacement for undefined table vales';
            $self->__opt_readline( $key, $prompt );
        }
        elsif ( $key eq 'progress_bar' ) {
            my $digits = 7;
            my $prompt = '"Threshold ProgressBar"';
            $self->__opt_number_range( $key, $prompt, $digits );
        }
        elsif ( $key eq 'max_rows' ) {
            my $digits = 7;
            my $prompt = '"Max rows"';
            $self->__opt_number_range( $key, $prompt, $digits );
        }
        elsif ( $key eq 'lock_stmt' ) {
            my $list = [ 'Lk0', 'Lk1' ];
            my $prompt = 'Keep statement';
            $self->__opt_choose_index( $key, $prompt, $list );
        }
        elsif ( $key eq 'metadata' ) {
            my $list = $self->{info}{yes_no};
            my $prompt = 'Enable Metadata';
            $self->__opt_choose_index( $key, $prompt, $list );
        }
        elsif ( $key eq 'regexp_case' ) {
            my $list = $self->{info}{yes_no};
            my $prompt = 'REGEXP case sensitiv';
            $self->__opt_choose_index( $key, $prompt, $list );
        }
        elsif ( $key eq '_parentheses' ) {
            my $sub_menu = $sub_menus->{$key};
            $self->__opt_choose_multi( $sub_menu );
        }
        elsif ( $key eq 'db_login_once' ) {
            my $list = [ 'per-DB', 'once' ];
            my $prompt = 'Ask for credentials';
            $self->__opt_choose_index( $key, $prompt, $list );
        }
        elsif ( $key eq '_env_dbi' ) {
            my $sub_menu = $sub_menus->{$key};
            $self->__opt_choose_multi( $sub_menu );
        }
        elsif ( $key eq 'db_defaults' ) {
            $self->database_setting();
        }
        elsif ( $key eq 'sssc_mode' ) {
            my $list = [ 'simple', 'compat' ];
            my $prompt = 'Sssc mode';
            $self->__opt_choose_index( $key, $prompt, $list );
        }
        elsif ( $key eq 'operators' ) {
            my $available = $self->{info}{avail_operators};
            $self->__opt_choose_a_list( $key, $available );
        }
        elsif ( $key eq 'db_drivers' ) {
            my $available = $self->{info}{avail_db_drivers};
            $self->__opt_choose_a_list( $key, $available );
        }
        elsif ( $key eq 'mouse' ) {
            my $max = 4;
            my $prompt = 'Mouse mode';
            $self->__opt_number( $key, $prompt, $max );
        }
        elsif ( $key eq '_enchant' ) {
            my $sub_menu = $sub_menus->{$key};
            $self->__opt_choose_multi( $sub_menu );
        }
        else { die "Unknown option: $key" }
    }
}


sub __opt_choose_multi {
    my ( $self, $sub_menu ) = @_;
    my $changed = choose_multi( $sub_menu, $self->{opt} );
    return if ! $changed;
    $self->{info}{write_config}++;
}


sub __opt_choose_index {
    my ( $self, $key, $prompt, $list ) = @_;
    my $yn = 0;
    my $current = $list->[$self->{opt}{$key}];
    # Choose
    my $idx = choose(
        [ undef, @$list ],
        { %{$self->{info}{lyt_1}}, prompt => $prompt . ' [' . $current . ']:', index => 1 }
    );
    return if ! defined $idx;
    return if $idx == 0;
    $idx--;
    $self->{opt}{$key} = $idx;
    $self->{info}{write_config}++;
    return;
}

sub __opt_choose_a_list {
    my ( $self, $key, $available ) = @_;
    my $current = $self->{opt}{$key};
    # Choose_list
    my $list = choose_a_subset( $available, { current => $current } );
    return if ! defined $list;
    $self->{opt}{$key} = $list;
    $self->{info}{write_config}++;
    return;
}

sub __opt_number {
    my ( $self, $key, $prompt, $max ) = @_;
    my $current = $self->{opt}{$key};
    # Choose
    my $choice = choose(
        [ undef, 0 .. $max ],
        { %{$self->{info}{lyt_1}}, prompt => $prompt . ' [' . $current . ']:', justify => 1 }
    );
    return if ! defined $choice;
    $self->{opt}{$key} = $choice;
    $self->{info}{write_config}++;
    return;
}

sub __opt_number_range {
    my ( $self, $key, $prompt, $digits ) = @_;
    my $current = $self->{opt}{$key};
    $current = insert_sep( $current, $self->{opt}{thsd_sep} );
    # Choose_a_number
    my $choice = choose_a_number( $digits, { name => $prompt, current => $current } );
    return if ! defined $choice;
    $self->{opt}{$key} = $choice eq '--' ? undef : $choice;
    $self->{info}{write_config}++;
    return;
}

sub __opt_readline {
    my ( $self, $key, $prompt ) = @_;
    my $current = $self->{opt}{$key};
    $prompt .= ' ["' . $current . '"]: ';
    # Readline
    my $choice = util_readline( $prompt );
    return if ! defined $choice;
    $self->{opt}{$key} = $choice;
    $self->{info}{write_config}++;
    return;
}


sub database_setting {
    my ( $self, $db ) = @_;
    my ( $db_driver, $section );
    if ( ! defined $db ) {
        if ( @{$self->{opt}{db_drivers}} == 1 ) {
            $db_driver = $self->{opt}{db_drivers}[0];
        }
        else {
            # Choose
            $db_driver = choose(
                [ undef, @{$self->{opt}{db_drivers}} ],
                { %{$self->{info}{lyt_1}} }
            );
            return if ! defined $db_driver;
        }
        $section = $db_driver;
    }
    else {
        $db_driver = $self->{info}{db_driver};
        $section   = $db_driver . '_' . $db;
        for my $key ( keys %{$self->{opt}{$db_driver}} ) {
            $self->{opt}{$section}{$key} //= $self->{opt}{$db_driver}{$key};
        }
    }

    my $orig = clone( $self->{opt} );
    my $menus = {
        SQLite => [
            [ 'sqlite_unicode',             "- Unicode" ],
            [ 'sqlite_see_if_its_a_number', "- See if its a number" ],
            [ '_binary_filter',             "- Binary Filter" ],
        ],
        mysql => [
            [ 'host',              "- Host" ],
            [ 'port',              "- Port" ],
            [ 'mysql_enable_utf8', "- Enable utf8" ],
            [ '_binary_filter',    "- Binary Filter" ],
        ],
        Pg => [
            [ 'host',           "- Host" ],
            [ 'port',           "- Port" ],
            [ 'pg_enable_utf8', "- Enable utf8" ],
            [ '_binary_filter', "- Binary Filter" ],
        ],
    };
    my $prompt;
    if ( defined $db ) {
        $prompt = 'DB: "' . ( $db_driver eq 'SQLite' ? basename( $db ) : $db ) . '"';
    }
    else {
        $prompt = 'Driver: ' . $db_driver;
    }
    my @pre = ( undef, $self->{info}{_confirm} );
    my @real = map { $_->[1] } @{$menus->{$db_driver}};
    my $choices = [ @pre, @real ];
    push @$choices, "  RESET" if defined $db;

    DB_OPTION: while ( 1 ) {
        # Choose
        my $idx = choose(
            $choices,
            { %{$self->{info}{lyt_3}}, index => 1, prompt => $prompt  }
        );
        exit if ! defined $idx;
        my $key;
        if ( $idx <= $#pre ) {
            $key = $pre[$idx];
        }
        elsif ( $idx == @pre + @real ) {
            for my $key ( keys %{$self->{opt}{$section}} ) {
                $self->{opt}{$section}{$key} = undef;
            }
            $self->{info}{write_config}++;
            next;
        }
        else {
            $idx -= @pre;
            $key = $menus->{$db_driver}[$idx][0];
            die if ! exists $self->{opt}{$db_driver}{$key};
        }
        if ( ! defined $key ) {
            if ( $self->{info}{write_config} ) {
                $self->{opt} = clone( $orig );
            }
            return;
        }
        if ( $key eq $self->{info}{_confirm} ) {
            if ( $self->{info}{write_config} ) {
                $self->__write_config_file( $self->{info}{config_file} );
                delete $self->{info}{write_config};
                return 1;
            }
            return;
        }
        if ( $db_driver eq "SQLite" ) {
            if ( $key eq 'sqlite_unicode' ) {
                my $list = $self->{info}{yes_no};
                my $prompt = 'Unicode';
                $self->__db_opt_choose_index( $section, $key, $prompt, $list );
            }
            elsif ( $key eq 'sqlite_see_if_its_a_number' ) {
                my $list = $self->{info}{yes_no};
                my $prompt = 'See if its a number';
                $self->__db_opt_choose_index( $section, $key, $prompt, $list );
            }
            elsif ( $key eq '_binary_filter' ) {
                my $list = $self->{info}{yes_no};
                my $prompt = 'Enable Binary Filter';
                $self->__db_opt_choose_index( $section, $key, $prompt, $list );
            }
            else { die "Unknown key: $key" }
        }
        elsif ( $db_driver eq "mysql" ) {
            if ( $key eq 'mysql_enable_utf8' ) {
                my $list = $self->{info}{yes_no};
                my $prompt = 'Enable utf8';
                $self->__db_opt_choose_index( $section, $key, $prompt, $list );
            }
            elsif ( $key eq 'host' ) {
                my $prompt = 'Host';
                $self->__db_opt_readline( $section, $key, $prompt );
            }
            elsif ( $key eq 'port' ) {
                my $prompt = 'Port';
                $self->__db_opt_readline( $section, $key, $prompt );
            }
            elsif ( $key eq '_binary_filter' ) {
                my $list = $self->{info}{yes_no};
                my $prompt = 'Enable Binary Filter';
                $self->__db_opt_choose_index( $section, $key, $prompt, $list );
            }
            else { die "Unknown key: $key" }
        }
        elsif ( $db_driver eq "Pg" ) {
            if ( $key eq 'pg_enable_utf8' ) {
                my $prompt = 'Enable utf8';
                my $list = [ @{$self->{info}{yes_no}}, 'AUTO' ];
                $self->__db_opt_choose_index( $section, $key, $prompt, $list );
                $self->{opt}{$section}{$key} = -1 if $self->{opt}{$section}{$key} == 2;
            }
            elsif ( $key eq 'host' ) {
                my $prompt = 'Host';
                $self->__db_opt_readline( $section, $key, $prompt );
            }
            elsif ( $key eq 'port' ) {
                my $prompt = 'Port';
                $self->__db_opt_readline( $section, $key, $prompt );
            }
            elsif ( $key eq '_binary_filter' ) {
                my $list = $self->{info}{yes_no};
                my $prompt = 'Enable Binary Filter';
                $self->__db_opt_choose_index( $section, $key, $prompt, $list );
            }
            else { die "Unknown key: $key" }
        }
    }
}


sub __db_opt_choose_index {
    my ( $self, $section, $key, $prompt, $list ) = @_;
    my $current = $list->[$self->{opt}{$section}{$key}];
    # Choose
    my $idx = choose(
        [ undef, @$list ],
        { %{$self->{info}{lyt_1}}, prompt => $prompt . ' [' . $current . ']:', index => 1 }
    );
    return if ! defined $idx;
    return if $idx == 0;
    $idx--;
    $self->{opt}{$section}{$key} = $idx;
    $self->{info}{write_config}++;
    return;
}


sub __db_opt_readline {
    my ( $self, $section, $key, $prompt ) = @_;
    my $current = $self->{opt}{$section}{$key};
    $prompt .= ' ["' . $current . '"]: ';
    # Readline
    my $choice = util_readline( $prompt );
    return if ! defined $choice;
    $self->{opt}{$section}{$key} = $choice;
    $self->{info}{write_config}++;
    return;
}


sub __write_config_file {
    my ( $self, $file ) = @_;
    my $tmp = {};
    for my $section ( sort keys %{$self->{opt}} ) {
        if ( ref( $self->{opt}{$section} ) eq 'HASH' ) {
            for my $key ( keys %{$self->{opt}{$section}} ) {
                $tmp->{$section}{$key} = $self->{opt}{$section}{$key};
            }
        }
        else {
            my $key = $section;
            my $section = $self->{info}{sect_generic};
            $tmp->{$section}{$key} = $self->{opt}{$key};
        }
    }
    $self->write_json( $file, $tmp );
}


sub read_config_file {
    my ( $self, $file ) = @_;
    my $tmp = $self->read_json( $file );
    for my $section ( keys %$tmp ) {
        for my $key ( keys %{$tmp->{$section}} ) {
            if ( $section eq $self->{info}{sect_generic} ) {
                $self->{opt}{$key} = $tmp->{$section}{$key};
            }
            else {
                $self->{opt}{$section}{$key} = $tmp->{$section}{$key};
            }
        }
    }
    return $self->{opt};
}


sub write_json {
    my ( $self, $file, $h_ref ) = @_;
    my $json = JSON::XS->new->pretty->canonical->encode( $h_ref );
    open my $fh, '>', encode( 'locale_fs', $file ) or die $!;
    print $fh $json;
    close $fh;
}


sub read_json {
    my ( $self, $file ) = @_;
    return {} if ! -f encode( 'locale_fs', $file );
    open my $fh, '<', encode( 'locale_fs', $file ) or die $!;
    my $json = do { local $/; <$fh> };
    close $fh;
    my $h_ref = {};
    $h_ref = decode_json( $json ) if $json;
    return $h_ref;
}


1;


__END__
